eval/topdown_test.go

```diff
@@ -457,6 +457,30 @@ func TestTopDownDisjunction(t *testing.T) {
 	}
 }
 
+func TestTopDownNegation(t *testing.T) {
+	tests := []struct {
+		note     string
+		rules    []string
+		expected interface{}
+	}{
+		{"neg: constants", []string{"p = true :- not true = false"}, "true"},
+		{"neg: constants", []string{"p = true :- not true = true"}, ""},
+		{"neg: array contains", []string{"p = true :- not a[_] = 9999"}, "true"},
+		{"neg: array diff", []string{"p = true :- not a[_] = 4"}, ""},
```

Comment:
I can't remember what we decided about how negation works.  Usually it would be unsafe to have a variable inside a negation that wasn't ground at the time of evaluating the negation.  So p[x] :- q[x], not r[x] is safe, but p[x] :- q[x], not r[y] is not safe.  For p=true :- not a[_] = 4 I'm not sure what the semantics are: is it asking if there is anything in a that = 4, or is it asking if everything in 'a' does not = 4.  At any rate, the usual approach is to avoid this case and require all vars to be ground before evaluating the negation, which you are doing below.  So if that's the semantics we're using then we should remove the tests with unsafe rules since people may think that's permitted.